contracts/utils/math/Math.sol

```diff
@@ -537,41 +537,101 @@ library Math {
+        //
+        // 2. Compute `n = x mod 255`, the `n` is used to split the equation in two parts that can be more efficiently
+        //    computed using the `Logarithm's Product Rule`:
+        //      - lemma: log2(x) == log2(x / n) + log2(n)
+        //    this `n` guarantees that `x / n` is a power of two multiple of 256, we use this fact to build a lookup table
+        //    that calculates `log2(x / n)` and `log2(n)` very efficiently.
+        //
+        // 3. Compute `prod0 = log2(n)`, given `x % 255` is a power of two, there are only 8 possible values for `n`, so
+        //    `log2(n)` can be easily computed using a lookup table, here we use the opcode `BYTE` to lookup a 32-byte word.
+        //
+        // 4. Compute `prod1 = log2(x / n)`, we use the fact that `x / n` is a power of two multiple of 256, so there's exactly
+        //    32 possible distinct values for `prod1`, respectively: 0, 8, 16, .. 240, 248. This allow an efficient lookup
+        //    table to be created using a single 32-byte word, we extract `log2(x/n)` from the table by compute:
+        //    prod1 = log2(x / n) = (x / n * table) >> 248
+        //
+        // 5. The final result is simply the sum of `prod0` and `prod1` calculated previously:
+        //    log2(x) = log2(x / n) + log2(n) = prod0 + prod1
+        //
+        // @author Lohann Ferreira <developer@lohann.dev>
         unchecked {
-            exp = 128 * SafeCast.toUint(value > (1 << 128) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 64 * SafeCast.toUint(value > (1 << 64) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 32 * SafeCast.toUint(value > (1 << 32) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 16 * SafeCast.toUint(value > (1 << 16) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 8 * SafeCast.toUint(value > (1 << 8) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 4 * SafeCast.toUint(value > (1 << 4) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 2 * SafeCast.toUint(value > (1 << 2) - 1);
-            value >>= exp;
-            result += exp;
+            // Round `x` down to the closest power of 2 using an modified version of Seander's algorithm.
+            // Reference: https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+            x |= x >> 1;
+            x |= x >> 2;
+            x |= x >> 4;
+            x |= x >> 8;
+            x |= x >> 16;
+            x |= x >> 32;
+            x |= x >> 64;
+            x |= x >> 128;
+            // note `x = 0` results in 1 here, when the closest power of two is actually zero given
+            // `2**-infinity == 0` then we should do `(x >> 1) + toUint(x > 0)` instead, but this is not
+            // necessary given `floor(log2(0)) == floor(log2(1))` anyway.
```

Comment:
```suggestion
            // necessary given `floor(log2(0)) == floor(log2(1))` anyway.
```
Strictly speaking, `floor(log2(0))` is minux infinity. I'm guessing were `log2` represent the outcome of this function, and not the mathematical log2. Maybe we should rephrase it a bit.

Reply:
True, once we are dealing with natural numbers the actual result is not representable, theoretically the correct would be revert like a division by zero, but then this would break a lot of libraries who rely on it.
Will think in a better description here.

Reply:
Not required anymore, once switch to @cairoeth implementation.